src/bson.c

```diff
@@ -86,59 +88,86 @@ char* php_phongo_field_path_as_string(php_phongo_field_path* field_path)
 	return path;
 }
 
-php_phongo_field_path* php_phongo_field_path_alloc(void)
+php_phongo_field_path* php_phongo_field_path_alloc(bool free_elements)
 {
 	php_phongo_field_path* tmp = ecalloc(1, sizeof(php_phongo_field_path));
 	tmp->ref_count             = 1;
+	tmp->free_elements         = free_elements;
 
 	return tmp;
 }
 
 void php_phongo_field_path_free(php_phongo_field_path* field_path)
 {
+	if (field_path->free_elements) {
+		size_t i;
+
+		for (i = 0; i < field_path->size; i++) {
+			efree(field_path->elements[i]);
+		}
+	}
 	if (field_path->elements) {
 		efree(field_path->elements);
 	}
+	if (field_path->element_types) {
+		efree(field_path->element_types);
+	}
 	efree(field_path);
 }
 
 static void php_phongo_field_path_ensure_allocation(php_phongo_field_path* field_path, size_t level)
 {
-	if (level >= field_path->allocated_levels) {
+	if (level >= field_path->allocated) {
 		size_t i;
 
-		field_path->allocated_levels = field_path->current_level + PHONGO_FIELD_PATH_EXPANSION;
-		field_path->elements         = erealloc(field_path->elements, sizeof(char**) * field_path->allocated_levels);
+		field_path->allocated     = field_path->size + PHONGO_FIELD_PATH_EXPANSION;
+		field_path->elements      = erealloc(field_path->elements, sizeof(char**) * field_path->allocated);
+		field_path->element_types = erealloc(field_path->element_types, sizeof(php_phongo_bson_field_path_item_types*) * field_path->allocated);
```

Comment:
Similarly, shouldn't this be `sizeof(php_phongo_bson_field_path_item_types)` as well, since `element_types` points to an array of enums?

Reply:
Probably - will have a look and update if needed.